src/main/java/com/google/changestreams/sample/bigquery/changelog/fullschema/SpannerChangeStreamsToBigQuery.java

```diff
@@ -0,0 +1,329 @@
+
+  /**
+   * Main entry point for executing the pipeline.
+   *
+   * @param args The command-line arguments to the pipeline.
+   */
+  public static void main(String[] args) {
+    LOG.info("Starting to replicate change records from Spanner Change Streams to BigQuery");
+
+    Options options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
+
+    options.setStreaming(true);
+    options.setEnableStreamingEngine(true);
+
+    validateOptions(options);
+    run(options).waitUntilFinish();
+  }
+
+  private static void validateOptions(Options options) {
+    final String startTimestampStr = options.getStartTimestamp();
+    final Timestamp startTimestamp = Timestamp.parseTimestamp(startTimestampStr);
+    final String endTimestampStr = options.getEndTimestamp();
+    if (!endTimestampStr.isEmpty()) {
+      final Timestamp endTimestamp = Timestamp.parseTimestamp(endTimestampStr);
+      if (startTimestamp.compareTo(endTimestamp) <= 0) {
+        throw new IllegalArgumentException(
+          "startTimestamp must be smaller than or equal to endTimestamp.");
+      }
+    }
+
+    if (options.getDlqRetryMinutes() <= 0) {
+      throw new IllegalArgumentException("dlqRetryMinutes must be positive.");
+    }
+  }
+
+  /**
+   * Runs the pipeline with the supplied options.
+   *
+   * @param options The execution parameters to the pipeline.
+   * @return The result of the pipeline execution.
+   */
+  public static PipelineResult run(Options options) {
+    /*
+     * Stages:
+     *   1) Read DataChangeRecord from Change Streams.
+     *   2) Create FailsafeElement of Mod JSON and merge from:
+     *       - DataChangeRecord.
+     *       - GCS Dead letter queue.
+     *   3) Convert Mod JSON into TableRow by reading from Spanner at Mod commit timestamp.
+     *   4) Append TableRow to BigQuery.
+     *   5) Write Failures from 2), 3) and 4) to GCS dead letter queue.
+     */
+
+    Pipeline pipeline = Pipeline.create(options);
+    DeadLetterQueueManager dlqManager = buildDlqManager(options);
+
+    String dlqDirectory = dlqManager.getRetryDlqDirectoryWithDateTime();
+    String tempDlqDirectory = dlqManager.getRetryDlqDirectory() + "tmp/";
+
+    Timestamp startTimestamp = Timestamp.parseTimestamp(options.getStartTimestamp());
+
+    /*
+     * Stage 1: Ingest and Normalize Data to FailsafeElement with JSON Strings
+     *   a) Read DataStream data from GCS into JSON String FailsafeElements (datastreamJsonRecords)
+     *   b) Reconsume Dead Letter Queue data from GCS into JSON String FailsafeElements
+     *     (dlqJsonRecords)
+     *   c) Flatten DataStream and DLQ Streams (jsonRecords)
+     */
+    SpannerIO.ReadChangeStream readChangeStream = SpannerIO
+      .readChangeStream()
+      .withSpannerConfig(
+        SpannerConfig
+          .create()
+          .withHost(ValueProvider.StaticValueProvider.of(options.getSpannerHost()))
```

Comment:
Do we want to add this line here? Will it confuse change stream users?

Reply:
I remember I saw it was in one of the template code, we can keep it but we don't have to document it?